File: aoc-2023/src/main/java/com/capital7software/aoc/aoc2023aoc/days/Day02.java
package com.capital7software.aoc.aoc2023aoc.days;

import com.capital7software.aoc.lib.AdventOfCodeSolution;
import com.capital7software.aoc.lib.analysis.CubeConundrum;
import java.time.Instant;
import java.util.List;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * --- Day 2: Cube Conundrum ---<br><br>
 * You're launched high into the atmosphere! The apex of your trajectory just barely reaches the
 * surface of a large island floating in the sky. You gently land in a fluffy pile of leaves.
 * It's quite cold, but you don't see much snow. An Elf runs over to greet you.
 *
 * <p><br>
 * The Elf explains that you've arrived at Snow Island and apologizes for the lack of snow.
 * He'll be happy to explain the situation, but it's a bit of a walk, so you have some time.
 * They don't get many visitors up here; would you like to play a game in the meantime?
 *
 * <p><br>
 * As you walk, the Elf shows you a small bag and some cubes which are either red, green, or blue.
 * Each time you play this game, he will hide a secret number of cubes of each color in the bag,
 * and your goal is to figure out information about the number of cubes.
 *
 * <p><br>
 * To get information, once a bag has been loaded with cubes, the Elf will reach into the bag,
 * grab a handful of random cubes, show them to you, and then put them back in the bag.
 * He'll do this a few times per game.
 *
 * <p><br>
 * You play several games and record the information from each game (your puzzle input).
 * Each game is listed with its ID number (like the 11 in Game 11: ...) followed by a
 * semicolon-separated list of subsets of cubes that were revealed from the bag
 * (like 3 red, 5 green, 4 blue).
 *
 * <p><br>
 * For example, the record of a few games might look like this:
 * <ul>
 *     <li>
 *         Game 1: 3 blue, 4 red; 1 red, 2 green, 6 blue; 2 green
 *     </li>
 *     <li>
 *         Game 2: 1 blue, 2 green; 3 green, 4 blue, 1 red; 1 green, 1 blue
 *     </li>
 *     <li>
 *         Game 3: 8 green, 6 blue, 20 red; 5 blue, 4 red, 13 green; 5 green, 1 red
 *     </li>
 *     <li>
 *         Game 4: 1 green, 3 red, 6 blue; 3 green, 6 red; 3 green, 15 blue, 14 red
 *     </li>
 *     <li>
 *         Game 5: 6 red, 1 blue, 3 green; 2 blue, 1 red, 2 green
 *     </li>
 * </ul>
 * In game 1, three sets of cubes are revealed from the bag (and then put back again).
 * The first set is 3 blue cubes and 4 red cubes; the second set is 1 red cube, 2 green cubes,
 * and 6 blue cubes; the third set is only 2 green cubes.
 *
 * <p><br>
 * The Elf would first like to know which games would have been possible if the bag contained
 * only 12 red cubes, 13 green cubes, and 14 blue cubes?
 *
 * <p><br>
 * In the example above, games 1, 2, and 5 would have been possible if the bag had been
 * loaded with that configuration. However, game 3 would have been impossible because at
 * one point the Elf showed you 20 red cubes at once; similarly, game 4 would also have
 * been impossible because the Elf showed you 15 blue cubes at once. If you add up the IDs
 * of the games that would have been possible, you get 8.
 *
 * <p><br>
 * Determine which games would have been possible if the bag had been loaded with only 12
 * red cubes, 13 green cubes, and 14 blue cubes. What is the sum of the IDs of those games?
 *
 * <p><br>
 * Your puzzle answer was 2101.
 *
 * <p><br>
 * --- Part Two ---<br><br>
 * The Elf says they've stopped producing snow because they aren't getting any water!
 * He isn't sure why the water stopped; however, he can show you how to get to the water
 * source to check it out for yourself. It's just up ahead!
 *
 * <p><br>
 * As you continue your walk, the Elf poses a second question: in each game you played,
 * what is the fewest number of cubes of each color that could have been in the bag to
 * make the game possible?
 *
 * <p><br>
 * Again consider the example games from earlier:
 * <ul>
 *     <li>
 *         Game 1: 3 blue, 4 red; 1 red, 2 green, 6 blue; 2 green
 *     </li>
 *     <li>
 *         Game 2: 1 blue, 2 green; 3 green, 4 blue, 1 red; 1 green, 1 blue
 *     </li>
 *     <li>
 *         Game 3: 8 green, 6 blue, 20 red; 5 blue, 4 red, 13 green; 5 green, 1 red
 *     </li>
 *     <li>
 *         Game 4: 1 green, 3 red, 6 blue; 3 green, 6 red; 3 green, 15 blue, 14 red
 *     </li>
 *     <li>
 *         Game 5: 6 red, 1 blue, 3 green; 2 blue, 1 red, 2 green
 *     </li>
 * </ul>
 * <ul>
 *     <li>
 *         In game 1, the game could have been played with as few as 4 red, 2 green,
 *         and 6 blue cubes. If any color had even one fewer cube, the game would have
 *         been impossible.
 *     </li>
 *     <li>
 *         Game 2 could have been played with a minimum of 1 red, 3 green, and 4 blue cubes.
 *     </li>
 *     <li>
 *         Game 3 must have been played with at least 20 red, 13 green, and 6 blue cubes.
 *     </li>
 *     <li>
 *         Game 4 required at least 14 red, 3 green, and 15 blue cubes.
 *     </li>
 *     <li>
 *         Game 5 needed no fewer than 6 red, 3 green, and 2 blue cubes in the bag.
 *     </li>
 * </ul>
 * <b>The power of a set of cubes is equal to the numbers of red, green, and blue cubes
 * multiplied together.</b> The power of the minimum set of cubes in game 1 is 48.
 * In games 2-5 it was 12, 1560, 630, and 36, respectively. Adding up these five powers
 * produces the sum 2286.
 *
 * <p><br>
 * For each game, find the minimum set of cubes that must have been present.
 * What is the sum of the power of these sets?
 *
 * <p><br>
 * Your puzzle answer was 58269.
 */
public class Day02 implements AdventOfCodeSolution {
  private static final Logger LOGGER = LoggerFactory.getLogger(Day02.class);

  /**
   * Instantiates the solution instance.
   */
  public Day02() {

  }

  @Override
  public String getDefaultInputFilename() {
    return "inputs/input_day_02-01.txt";
  }

  @Override
  public void runPart1(List<String> input) {
    var start = Instant.now();
    var sum = getSumOfGameResultIds(input);
    var end = Instant.now();

    LOGGER.info("The sum of the possible Game IDs is: {}", sum);
    logTimings(LOGGER, start, end);
  }

  @Override
  public void runPart2(List<String> input) {
    var start = Instant.now();
    var sum = getSumOfGameResultPower(input);
    var end = Instant.now();

    LOGGER.info("The sum of the Game Powers is: {}", sum);
    logTimings(LOGGER, start, end);
  }

  /**
   * Returns the sum of the GameResult IDs that are possible.
   *
   * @param input The List of games that have been played.
   * @return The sum of the GameResult IDs that are possible!
   */
  public int getSumOfGameResultIds(List<String> input) {
    var game = new CubeConundrum();

    return game.getGameResults(input)
        .stream()
        .filter(CubeConundrum.GameResult::isPossible)
        .mapToInt(CubeConundrum.GameResult::id)
        .sum();
  }

  /**
   * Returns the sum of the Power of the GameResult IDs.
   *
   * @param input The List of games that have been played.
   * @return The sum of the Power of the GameResult IDs!
   */
  public int getSumOfGameResultPower(List<String> input) {
    var game = new CubeConundrum();

    return game.getGameResults(input)
        .stream()
        .mapToInt(CubeConundrum.GameResult::power)
        .sum();
  }
}

File: aoc-2017/src/main/kotlin/com/capital7software/aoc/aoc2017aoc/days/Day02.kt
package com.capital7software.aoc.aoc2017aoc.days

import com.capital7software.aoc.lib.AdventOfCodeSolution
import com.capital7software.aoc.lib.string.CorruptionChecksum
import java.time.Instant
import org.slf4j.Logger
import org.slf4j.LoggerFactory

/**
 * **--- Day 2: Corruption Checksum ---**
 *
 * As you walk through the door, a glowing humanoid shape yells in your direction.
 * "You there! Your state appears to be idle. Come help us repair the corruption in
 * this spreadsheet - if we take another millisecond, we'll have to display an hourglass cursor!"
 *
 * The spreadsheet consists of rows of apparently-random numbers. To make sure the
 * recovery process is on the right track, they need you to calculate the spreadsheet's
 * **checksum**. For each row, determine the difference between the largest value and the
 * smallest value; the checksum is the sum of all of these differences.
 *
 * For example, given the following spreadsheet:
 *
 * ```
 * 5 1 9 5
 * 7 5 3
 * 2 4 6 8
 * ```
 *
 * - The first row's largest and smallest values are 9 and 1, and their difference is 8.
 * - The second row's largest and smallest values are 7 and 3, and their difference is 4.
 * - The third row's difference is 6.
 *
 * In this example, the spreadsheet's checksum would be 8 + 4 + 6 = 18.
 *
 * **What is the checksum** for the spreadsheet in your puzzle input?
 *
 * Your puzzle answer was **32020**.
 *
 * **--- Part Two ---**
 *
 * "Great work; looks like we're on the right track after all. Here's a star for your effort."
 * However, the program seems a little worried. Can programs **be** worried?
 *
 * "Based on what we're seeing, it looks like all the User wanted is some information about
 * the **evenly divisible values** in the spreadsheet. Unfortunately, none of us are
 * equipped for that kind of calculation - most of us specialize in bitwise operations."
 *
 * It sounds like the goal is to find the only two numbers in each row where one evenly
 * divides the other - that is, where the result of the division operation is a whole
 * number. They would like you to find those numbers on each line, divide them,
 * and add up each line's result.
 *
 * For example, given the following spreadsheet:
 *
 * ```
 * 5 9 2 8
 * 9 4 7 3
 * 3 8 6 5
 * ```
 *
 * - In the first row, the only two numbers that evenly divide are 8 and 2; the result of this division is 4.
 * - In the second row, the two numbers are 9 and 3; the result is 3.
 * - In the third row, the result is 2.
 *
 * In this example, the sum of the results would be 4 + 3 + 2 = 9.
 *
 * What is the **sum of each row's result** in your puzzle input?
 *
 * Your puzzle answer was **236**.
 */
class Day02: AdventOfCodeSolution {
  companion object {
    private val log: Logger = LoggerFactory.getLogger(Day02::class.java)
  }

  override fun getDefaultInputFilename(): String = "inputs/input_day_02-01.txt"

  override fun runPart1(input: MutableList<String>) {
    val start = Instant.now()
    val answer = getChecksum(input)
    val end = Instant.now()

    log.info("$answer is the checksum for the spreadsheet rows!")
    logTimings(log, start, end)
  }

  override fun runPart2(input: MutableList<String>) {
    val start = Instant.now()
    val answer = getChecksum(input, true)
    val end = Instant.now()

    log.info("$answer is the checksum for the spreadsheet rows with evenly divisible numbers!")
    logTimings(log, start, end)
  }

  /**
   * Calculates the checksum for the specified list of spreadsheet rows.
   *
   * @param input The list of tab separated spreadsheet rows.
   * @param evenlyDivisible If true, then only the two numbers in each row that even divide one
   * another are used for the checksum.
   * @return An integer that is the checksum of the specified rows.
   */
  fun getChecksum(input: List<String>, evenlyDivisible: Boolean = false): Int {
    val instance = CorruptionChecksum(input)
    return instance.checksum(evenlyDivisible)
  }
}
